src/common/async.cpp

```diff
@@ -46,6 +49,13 @@ void Async::join() {
     }
 }
 
+Async::~Async() {
+    // WARNING: This MUST be here to make sure we break the loop before we
+    // stop the thread. Not doing that leads to undefined behavior.
+    break_loop();
+    join();
```

Comment:
1. Should you check if it's running before calling break_loop()? 
2. Why do you need to call join here, and witch thread are you blocking?